builder/publishCommand.go

```diff
@@ -0,0 +1,119 @@
+package builder
+
+import (
+	"assignment-exec/image-builder/constants"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/pkg/errors"
+	"io"
+	"log"
+	"os"
+)
+
+type publishCommand struct {
+	assgnEnv *assignmentEnv
+}
+
+func (cmd *publishCommand) execute() error {
+	return cmd.assgnEnv.publish()
+}
+
+func (cmd *publishCommand) undo() error {
+	err := cmd.assgnEnv.undoPublish()
+	if err != nil {
+		return errors.Wrap(err, "error in undo publish operation")
+	}
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) publish() error {
+	if assgnEnv.DockerfileData.Len() <= 0 {
```

Comment:
As mentioned above, in case where the language image is already present on hub and there are no other dependencies to be added. So in such case, only pull the image

Reply:
refer to [above comment](https://github.com/assignment-exec/image-builder/pull/20#discussion_r423293345).
Also, this function is responsible for publishing images so I don't see why this function should pull an image.

Reply:
Publish is the final phase, where based on the validations and dockerfile written, the image will be either pushed to hub or pulled from hub. I will change the function name if it is misleading.

Reply:
shouldn't the image already exist locally if we're at the publish phase?

Reply:
For example: if user has given configuration for language: gcc and version: 7. Then the image builder will first validate the base code-runner image provided, then it will check is the language image already exists, if no then it will write the dockerfile and build the image then it will move on to publish phase to either push the built image or pull the existing image.

Reply:
But I got it. I should instead pull the image in build phase only. Will change that